contracts/token/erc721/abstract/ERC721HybridPermit.sol

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
+import '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
+import "@openzeppelin/contracts/interfaces/IERC1271.sol";
+import "solidity-bytes-utils/contracts/BytesLib.sol";
+import "./IERC4494.sol";
+import "./ERC721Hybrid.sol";
+import "hardhat/console.sol";
+
+/**
+ * @title ERC721HybridPermit: An extension of the ERC721Hybrid NFT standard that supports off-chain approval via permits.
+ * @dev This contract implements ERC-4494 as well, allowing tokens to be approved via off-chain signed messages.
+ */
+
+abstract contract ERC721HybridPermit is ERC721Hybrid, IERC4494, EIP712 {
+
+    /** @notice mapping used to keep track of nonces of each token ID for validating
+     *  signatures
+     */
+    mapping(uint256 => uint256) private _nonces;
+
+    
+    /** @dev the unique identifier for the permit struct to be EIP 712 compliant */
+    bytes32 private constant _PERMIT_TYPEHASH = keccak256(
+        abi.encodePacked(
+            "Permit(",
+                "address spender,"
+                "uint256 tokenId,"
+                "uint256 nonce,"
+                "uint256 deadline"
+            ")"
+        )
+    );
+
+    constructor(string memory name, string memory symbol)
+        ERC721Hybrid(name, symbol)
+        EIP712(name, "1")
+    {}
+
+    /**
+     * @notice Function to approve by way of owner signature
+     * @param spender the address to approve
+     * @param tokenId the index of the NFT to approve the spender on
+     * @param deadline a timestamp expiry for the permit
+     * @param sig a traditional or EIP-2098 signature
+     */
+    function permit(
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        bytes memory sig
+    ) external override {
+        _permit(spender, tokenId, deadline, sig);
+    }
+
+    function _permit(
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        bytes memory sig
+    ) internal virtual {
+        if (deadline < block.timestamp) {
+            revert PermitExpired();
+        }
+
+        bytes32 digest = _buildPermitDigest(spender, tokenId, deadline);
+        address recoveredSigner;
+
+        if (sig.length == 64) {
+            // ERC2098 Sig
+            recoveredSigner = ECDSA.recover(
+                digest,
+                bytes32(BytesLib.slice(sig, 0, 32)),
+                bytes32(BytesLib.slice(sig, 32, 64))
+            );
+        } else if (sig.length == 65) {
+            // typical EDCSA Sig
+            recoveredSigner = ECDSA.recover(digest, sig);
+        } else {
+            revert InvalidSignature();
+        }
```

Comment:
**Question:** Would this `revert` for an `ERC-1271` based signature? My understanding is `ERC-1271` makes no assumptions about signatures and allows arbitrary length.

Reply:
Yep, moved the eip1271 validation to the top of the method. We still need to pass the digest and sig to the eip1271 wallet as thats the specified interface.